### PR TITLE
Wrap responses in result type

### LINKED ISSUE #9: Document http_get & collect_response must be in separate transactions

### LINKED ISSUE #21: shim http response
Put a `response` shim around http response to differentiate between status of the request & the remote server's response

```sql
-----------
-- TYPES --
-----------

-- A response from an HTTP server (abbreviated)
create type http_response AS (
    status_code integer,
    headers json,
    content text
);

-- Potential states of an HTTP Request
create type request_status as enum ('PENDING', 'SUCCESS', 'ERROR');

-- State wrapper around (any protocol) responses
create type net_result as (status request_status, response http_response, msg text);

--------------
-- Examples --
--------------

-- A (non-http) error has occured e.g. timeout
select ('ERROR', null, 'client side timeout occurred')::net_result;
select ('ERROR', null, 'invalid protocol')::net_result;

-- Has not been processed
select ('PENDING', null, 'request has not been processed')::net_result;

-- Successful request resulting in http 404
select ('SUCCESS', (1, 404, 'Not Found')::http_response, 'ok')::net_result;

-- Successful request resulting in http 200
select ('SUCCESS', (1, 200, 'Some data')::http_response, 'ok')::net_result;
```

### PR DESCRIPTION
**Pull request by olirice**
## What kind of change does this PR introduce?

Wraps http responses in a result type, making it possible to differentiate between the "request's" status/error/state and the the remote http server's response/status/state.

```
-----------
-- TYPES --
-----------

- Lifecycle states of a request (all protocols)
-- API: Public
create type net.request_status as enum ('PENDING', 'SUCCESS', 'ERROR');

-- A response from an HTTP server
-- API: Public
create type net.http_response AS (
    status_code integer,
    headers jsonb,
    content text
);

-- State wrapper around responses
-- API: Public
create type net.http_response_result as (
    status net.request_status,
    message text,
    response net.http_response
);

--------------
-- Examples --
--------------
select ('ERROR', 'invalid protocol', null)::net.http_response_result;

select (
        'SUCCESS',
        'ok',
        (
            404,
            '{"Content-Type": "application/json"}',
            '{"msg": "not found"}'
        )::net.http_response
    )::net.http_response_result;
```

The updated API docs are available at https://supabase.github.io/pg_net/api/ and a screenshot is below

resolves #21 
resolves #9

Screen Shot 2021-07-27 at 6 06 09 AM

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -50,18 +59,31 @@ net.http_collect_response(
...
---8<-- "test/expected/test_http_collect_response.out"
-```
+select net.http_get('https://news.ycombinator.com') as request_id;
+request_id
+----------
+         1
+(1 row)
 
-## SMTP
+select * from net.http_collect_response(1, async:=False);
```
Comment (kiwicopple) [inline docs/api.md:77]: `async:=false` (lowercase for consitency?)

File: test/test_http_get_collect.py [hunk 2]
```diff
@@ -43,11 +61,28 @@ def test_http_get_collect_async_false(sess):
     response = sess.execute(
         text(
             """
-        select * from net.http_collect_response(:request_id, async:=False); 
+        select * from net.http_collect_response(:request_id, async:=True);
```
Comment (kiwicopple) [inline test/test_http_get_collect.py:64]: `async:=true`  - perhaps there is a reason for the uppercase though?

**Review by kiwicopple — APPROVED** [review]
👍 nice work Oli

File: test/test_http_get_collect.py (hunk 2 above)
Comment (olirice) [reply test/test_http_get_collect.py:64]: python muscle memory

**Pull request merged**